cmd/ping.go

```diff
@@ -0,0 +1,240 @@
+	statusJSONBytes, e := json.MarshalIndent(pr, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+
+	return string(statusJSONBytes)
+}
+
+// String colorized ping result message.
+func (pr PingResult) String() (msg string) {
+	console.SetColor("Info", color.New(color.FgGreen, color.Bold))
+	console.SetColor("InfoFail", color.New(color.FgRed, color.Bold))
+	for _, ping := range pr.Pings {
+		if ping.Error == "" {
+			coloredDot := console.Colorize("Info", dot)
+			// Print server title
+			msg += fmt.Sprintf("%s  %s", coloredDot, console.Colorize("PrintB", ping.Server))
+			msg += fmt.Sprintf(" => %s\n", ping.Latency)
+		} else {
+			coloredDot := console.Colorize("InfoFail", dot)
+			msg += fmt.Sprintf("%s  %s", coloredDot, console.Colorize("PrintB", ping.Server))
+			msg += fmt.Sprintf(" => Error: %s\n", console.Colorize("InfoFail", ping.Error))
+		}
+	}
+	return msg
+}
+
+func fetchAdminInfo(alias, url string) madmin.InfoMessage {
+	// Create a new MinIO Admin Client
+	client, err := newAdminClient(alias)
+	fatalIf(err, "Unable to initialize admin connection.")
+
+	// Fetch info of all servers (cluster or single server)
+	admInfo, er := client.ServerInfo(globalContext)
+	// Keeps on retring until the server is up
+	for er != nil {
+		var ping Ping
+		var pings []Ping
+		ping.Server = url
+		ping.Error = er.Error()
+		pings = append(pings, ping)
+		time.Sleep(pingInterval)
+		printMsg(PingResult{Pings: pings})
+		admInfo, er = client.ServerInfo(globalContext)
+	}
+	return admInfo
+}
+
+// mainPing is entry point for ping command.
+func mainPing(ctx *cli.Context) error {
+	// check 'ping' cli arguments.
+	checkPingSyntax(ctx)
+
+	// Get the alias parameter from cli
+	args := ctx.Args()
+	alias := cleanAlias(args.Get(0))
+	hostConfig := mustGetHostConfig(alias)
+	if hostConfig == nil {
+		fatalIf(errInvalidAliasedURL(alias), "No such alias `"+alias+"` found.")
+		return nil
+	}
+
+	count := ctx.Int("count")
+	if count < 4 {
+		fatalIf(errInvalidArgument().Trace(ctx.Args()...), "please set count value more than 4")
+	}
+
+	httpClient := httpClient(10 * time.Second)
+
+	u, e := url.Parse(hostConfig.URL)
+	fatalIf(probe.NewError(e), "Unable to parse url")
+	admInfo := fetchAdminInfo(alias, hostConfig.URL)
+	for i := 0; i < count; i++ {
+		var pings []Ping
+		for _, server := range admInfo.Servers {
+			var ping Ping
+			req, e := http.NewRequest(http.MethodGet, u.Scheme+"://"+server.Endpoint+livelinessEndPoint, nil)
+			if e != nil {
+				return e
+			}
+			start := time.Now()
+			resp, e := httpClient.Do(req)
+			latency := computeLatency(start)
+			ping.Server = getServer(admInfo, server.Endpoint)
+			if e != nil {
+				ping.Error = e.Error()
+			}
+
+			if resp != nil {
+				switch resp.StatusCode {
+				case http.StatusOK:
+					ping.Latency = latency
+					ping.Error = ""
+					defer resp.Body.Close()
+				default:
+					ping.Error = resp.Status
+				}
+			}
+			pings = append(pings, ping)
```

Comment:
It would be much nicer if each failed attempt wrote the result, instead of waiting for all to complete.

Reply:
The reason for this is clubbing the o/p of all the servers of the cluster and displaying it together.

Reply:
@sinhaashish Sorry, could you rephrase that, I don't really understand it?

Reply:
As per the code we are printing `PingResult` which is the slice of `Ping` . This way we are clubbing the o/p of all the server and displaying . 
```
●  help.com:9000 => 809 μs
●  help.com:9001 => 781 μs
●  help.com:9002 => Error: Get "http://127.0.0.1:9002/minio/health/live": dial tcp 127.0.0.1:9002: connect: connection refused
●  help.com:9003 => 555 μs
```

If we instantly ping the error without waiting for other server o/p. then the servers will not be displayed together . I will be somewhat like 
```
●  help.com:9002 => Error: Get "http://127.0.0.1:9002/minio/health/live": dial tcp 127.0.0.1:9002: connect: connection refused

●  help.com:9000 => 809 μs
●  help.com:9001 => 781 μs
●  help.com:9003 => 555 μs
``` 

I thought to give the clubbed o/p of all server , that the reason i wait for response of all the servers.